src/main/java/com/harragan/battleshipsboot/service/GameArenaService.java

```diff
@@ -0,0 +1,151 @@
+        for (final Ship gameArenaShip : gameArena.getShipsOnBoard()) {
+            final boolean isShipOccupyingPosition =
+                    gameArenaShip.getOccupiedBoardPositions().stream()
+                            .anyMatch(ship.getOccupiedBoardPositions()::contains);
+
+            if (isShipOccupyingPosition) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public void registerHit(final BoardPosition boardPosition, final GameArena gameArena) {
+        boardPosition.setHit();
+        for (final Ship ship : gameArena.getShipsOnBoard()) {
+            if (ship.getOccupiedBoardPositions().contains(boardPosition)) {
+                setHitPositionOnShip(boardPosition, ship);
+                if (ship.isSunk()) {
+                    gameArena.addSunkenShip(ship);
+                }
+            }
+            gameArena.addShotPosition(boardPosition);
+        }
+    }
+
+    public BoardPosition getOccupiedPositionsOfShip(final BoardPosition boardPosition, final Ship ship) {
+        final Optional<BoardPosition> occupiedPosition =
+                ship.getOccupiedBoardPositions().stream().filter(p -> p.equals(boardPosition)).findFirst();
+
+        if (occupiedPosition.isPresent()) {
+            return occupiedPosition.get();
+        }
+        return null;
+    }
+
+    public void setOccupiedPositionsOfShip(final Ship ship) {
+        if (ship.getOrient() == Orientation.VERTICAL) {
+            for (int i = 1; i < ship.getLength(); i++) {
+                final BoardPosition pos =
+                        BoardPositionFactory.createBoardPosition(
+                                ship.getOccupiedPosition(0).getCol(), ship.getOccupiedPosition(0).getRow() + i);
+                final List<BoardPosition> occupiedBoardPositions = ship.getOccupiedBoardPositions();
+                occupiedBoardPositions.add(i, pos);
+                ship.setOccupiedPosition(occupiedBoardPositions);
+            }
+        } else {
+            for (int i = 1; i < ship.getLength(); i++) {
+                final char inputCol = ship.getOccupiedPosition(0).getCol();
+                final int input = inputCol + i;
+                final BoardPosition pos =
+                        BoardPositionFactory.createBoardPosition(
+                                (char) input, ship.getOccupiedPosition(0).getRow());
+                final List<BoardPosition> occupiedBoardPositions = ship.getOccupiedBoardPositions();
+                occupiedBoardPositions.add(i, pos);
+                ship.setOccupiedPosition(occupiedBoardPositions);
+            }
+        }
+    }
+
+    public void setHitPositionOnShip(final BoardPosition boardPosition, final Ship ship) {
+        for (final BoardPosition occupiedBoardPosition : ship.getOccupiedBoardPositions()) {
+            if (occupiedBoardPosition.equals(boardPosition)) {
+                occupiedBoardPosition.setHit();
```

Comment:
since there is only one possible hit at any one time you could short circuit the logic here and return after finding a hit

Reply:
Good point, thanks